{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8d7ec28-a607-4654-b8c8-5b27787b1cce",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"/home/bosser/anaconda3/lib/python3.12/runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/asyncio/base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/asyncio/base_events.py\", line 1986, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/asyncio/events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_50378/939138801.py\", line 1, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 26, in <module>\n",
      "    from pandas.compat import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/compat/__init__.py\", line 27, in <module>\n",
      "    from pandas.compat.pyarrow import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/compat/pyarrow.py\", line 8, in <module>\n",
      "    import pyarrow as pa\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n"
     ]
    },
    {
     "ename": "ImportError",
     "evalue": "\nA module that was compiled using NumPy 1.x cannot be run in\nNumPy 2.2.3 as it may crash. To support both 1.x and 2.x\nversions of NumPy, modules must be compiled with NumPy 2.0.\nSome module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n\nIf you are a user of the module, the easiest solution will be to\ndowngrade to 'numpy<2' or try to upgrade the affected module.\nWe expect that some modules will need time to support NumPy 2.\n\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py:44\u001b[0m, in \u001b[0;36m__getattr__\u001b[0;34m(attr_name)\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[38;5;66;03m# Also print the message (with traceback).  This is because old versions\u001b[39;00m\n\u001b[1;32m     40\u001b[0m     \u001b[38;5;66;03m# of NumPy unfortunately set up the import to replace (and hide) the\u001b[39;00m\n\u001b[1;32m     41\u001b[0m     \u001b[38;5;66;03m# error.  The traceback shouldn't be needed, but e.g. pytest plugins\u001b[39;00m\n\u001b[1;32m     42\u001b[0m     \u001b[38;5;66;03m# seem to swallow it and we should be failing anyway...\u001b[39;00m\n\u001b[1;32m     43\u001b[0m     sys\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mwrite(msg \u001b[38;5;241m+\u001b[39m tb_msg)\n\u001b[0;32m---> 44\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mImportError\u001b[39;00m(msg)\n\u001b[1;32m     46\u001b[0m ret \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(_multiarray_umath, attr_name, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m     47\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m ret \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[0;31mImportError\u001b[0m: \nA module that was compiled using NumPy 1.x cannot be run in\nNumPy 2.2.3 as it may crash. To support both 1.x and 2.x\nversions of NumPy, modules must be compiled with NumPy 2.0.\nSome module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n\nIf you are a user of the module, the easiest solution will be to\ndowngrade to 'numpy<2' or try to upgrade the affected module.\nWe expect that some modules will need time to support NumPy 2.\n\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"/home/bosser/anaconda3/lib/python3.12/runpy.py\", line 198, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/runpy.py\", line 88, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel_launcher.py\", line 18, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/traitlets/config/application.py\", line 1075, in launch_instance\n",
      "    app.start()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/kernelapp.py\", line 739, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/tornado/platform/asyncio.py\", line 205, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/asyncio/base_events.py\", line 641, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/asyncio/base_events.py\", line 1986, in _run_once\n",
      "    handle._run()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/asyncio/events.py\", line 88, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 545, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 534, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 437, in dispatch_shell\n",
      "    await result\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/ipkernel.py\", line 362, in execute_request\n",
      "    await super().execute_request(stream, ident, parent)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/kernelbase.py\", line 778, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/ipkernel.py\", line 449, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/ipykernel/zmqshell.py\", line 549, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3075, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3130, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/IPython/core/async_helpers.py\", line 128, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3334, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3517, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/IPython/core/interactiveshell.py\", line 3577, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"/tmp/ipykernel_50378/939138801.py\", line 1, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 49, in <module>\n",
      "    from pandas.core.api import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/core/api.py\", line 9, in <module>\n",
      "    from pandas.core.dtypes.dtypes import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/core/dtypes/dtypes.py\", line 24, in <module>\n",
      "    from pandas._libs import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n"
     ]
    },
    {
     "ename": "ImportError",
     "evalue": "\nA module that was compiled using NumPy 1.x cannot be run in\nNumPy 2.2.3 as it may crash. To support both 1.x and 2.x\nversions of NumPy, modules must be compiled with NumPy 2.0.\nSome module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n\nIf you are a user of the module, the easiest solution will be to\ndowngrade to 'numpy<2' or try to upgrade the affected module.\nWe expect that some modules will need time to support NumPy 2.\n\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py:44\u001b[0m, in \u001b[0;36m__getattr__\u001b[0;34m(attr_name)\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[38;5;66;03m# Also print the message (with traceback).  This is because old versions\u001b[39;00m\n\u001b[1;32m     40\u001b[0m     \u001b[38;5;66;03m# of NumPy unfortunately set up the import to replace (and hide) the\u001b[39;00m\n\u001b[1;32m     41\u001b[0m     \u001b[38;5;66;03m# error.  The traceback shouldn't be needed, but e.g. pytest plugins\u001b[39;00m\n\u001b[1;32m     42\u001b[0m     \u001b[38;5;66;03m# seem to swallow it and we should be failing anyway...\u001b[39;00m\n\u001b[1;32m     43\u001b[0m     sys\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mwrite(msg \u001b[38;5;241m+\u001b[39m tb_msg)\n\u001b[0;32m---> 44\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mImportError\u001b[39;00m(msg)\n\u001b[1;32m     46\u001b[0m ret \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(_multiarray_umath, attr_name, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[1;32m     47\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m ret \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[0;31mImportError\u001b[0m: \nA module that was compiled using NumPy 1.x cannot be run in\nNumPy 2.2.3 as it may crash. To support both 1.x and 2.x\nversions of NumPy, modules must be compiled with NumPy 2.0.\nSome module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n\nIf you are a user of the module, the easiest solution will be to\ndowngrade to 'numpy<2' or try to upgrade the affected module.\nWe expect that some modules will need time to support NumPy 2.\n\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import mysql.connector\n",
    "from unidecode import unidecode\n",
    "import joblib\n",
    "from sklearn.impute import KNNImputer\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.model_selection import GridSearchCV, train_test_split\n",
    "from sklearn.metrics import mean_squared_error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fc7fd1b0-3963-4b81-9039-d36f9fa8ee87",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_50378/2036770888.py:116: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.\n",
      "  df_general = pd.read_sql(query, db)\n",
      "/tmp/ipykernel_50378/2036770888.py:118: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.\n",
      "  top10tec_df = pd.read_sql(query_top10tec, db)\n",
      "/tmp/ipykernel_50378/2036770888.py:120: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.\n",
      "  top5hab_df = pd.read_sql(query_top5hab, db)\n"
     ]
    }
   ],
   "source": [
    "#Queries para obtener el df general + top 10 tecnologias y top 5 habilidades.\n",
    "\n",
    "config = toml.load(\"../.streamlit/secrets.toml\")\n",
    "db_config = config[\"database\"]\n",
    "\n",
    "db = mysql.connector.connect(host=db_config[\"host\"],\n",
    "                             user=db_config[\"user\"],\n",
    "                             password=db_config[\"password\"],\n",
    "                             database=db_config[\"database\"]) \n",
    "query = \"\"\"\n",
    "WITH TopTecnologias AS (\n",
    "    SELECT tr.tec_id, t.tecnologia\n",
    "    FROM tecnologias_relacion tr\n",
    "    JOIN tecnologias t ON tr.tec_id = t.tec_id\n",
    "    GROUP BY tr.tec_id\n",
    "    ORDER BY COUNT(*) DESC\n",
    "    LIMIT 10),\n",
    "TopHabilidades AS (\n",
    "    SELECT hr.hab_id, h.habilidad\n",
    "    FROM habilidades_relacion hr\n",
    "    JOIN habilidades h ON hr.hab_id = h.hab_id\n",
    "    GROUP BY hr.hab_id\n",
    "    ORDER BY COUNT(*) DESC\n",
    "    LIMIT 5),\n",
    "TecnologiasOferta AS (\n",
    "    SELECT \n",
    "        tr.id_oferta,\n",
    "        MAX(CASE WHEN tr.tec_id = (SELECT tec_id FROM TopTecnologias LIMIT 1 OFFSET 0) THEN 1 ELSE 0 END) AS `Tecnologia_1`,\n",
    "        MAX(CASE WHEN tr.tec_id = (SELECT tec_id FROM TopTecnologias LIMIT 1 OFFSET 1) THEN 1 ELSE 0 END) AS `Tecnologia_2`,\n",
    "        MAX(CASE WHEN tr.tec_id = (SELECT tec_id FROM TopTecnologias LIMIT 1 OFFSET 2) THEN 1 ELSE 0 END) AS `Tecnologia_3`,\n",
    "        MAX(CASE WHEN tr.tec_id = (SELECT tec_id FROM TopTecnologias LIMIT 1 OFFSET 3) THEN 1 ELSE 0 END) AS `Tecnologia_4`,\n",
    "        MAX(CASE WHEN tr.tec_id = (SELECT tec_id FROM TopTecnologias LIMIT 1 OFFSET 4) THEN 1 ELSE 0 END) AS `Tecnologia_5`,\n",
    "        MAX(CASE WHEN tr.tec_id = (SELECT tec_id FROM TopTecnologias LIMIT 1 OFFSET 5) THEN 1 ELSE 0 END) AS `Tecnologia_6`,\n",
    "        MAX(CASE WHEN tr.tec_id = (SELECT tec_id FROM TopTecnologias LIMIT 1 OFFSET 6) THEN 1 ELSE 0 END) AS `Tecnologia_7`,\n",
    "        MAX(CASE WHEN tr.tec_id = (SELECT tec_id FROM TopTecnologias LIMIT 1 OFFSET 7) THEN 1 ELSE 0 END) AS `Tecnologia_8`,\n",
    "        MAX(CASE WHEN tr.tec_id = (SELECT tec_id FROM TopTecnologias LIMIT 1 OFFSET 8) THEN 1 ELSE 0 END) AS `Tecnologia_9`,\n",
    "        MAX(CASE WHEN tr.tec_id = (SELECT tec_id FROM TopTecnologias LIMIT 1 OFFSET 9) THEN 1 ELSE 0 END) AS `Tecnologia_10`\n",
    "    FROM tecnologias_relacion tr\n",
    "    GROUP BY tr.id_oferta),\n",
    "HabilidadesOferta AS (\n",
    "    SELECT \n",
    "        hr.id_oferta,\n",
    "        MAX(CASE WHEN hr.hab_id = (SELECT hab_id FROM TopHabilidades LIMIT 1 OFFSET 0) THEN 1 ELSE 0 END) AS `Habilidad_1`,\n",
    "        MAX(CASE WHEN hr.hab_id = (SELECT hab_id FROM TopHabilidades LIMIT 1 OFFSET 1) THEN 1 ELSE 0 END) AS `Habilidad_2`,\n",
    "        MAX(CASE WHEN hr.hab_id = (SELECT hab_id FROM TopHabilidades LIMIT 1 OFFSET 2) THEN 1 ELSE 0 END) AS `Habilidad_3`,\n",
    "        MAX(CASE WHEN hr.hab_id = (SELECT hab_id FROM TopHabilidades LIMIT 1 OFFSET 3) THEN 1 ELSE 0 END) AS `Habilidad_4`,\n",
    "        MAX(CASE WHEN hr.hab_id = (SELECT hab_id FROM TopHabilidades LIMIT 1 OFFSET 4) THEN 1 ELSE 0 END) AS `Habilidad_5`\n",
    "    FROM habilidades_relacion hr\n",
    "    GROUP BY hr.id_oferta),\n",
    "PrimerIdioma AS (\n",
    "    SELECT \n",
    "        id_oferta,\n",
    "        MIN(idioma) AS PrimerIdioma\n",
    "    FROM idiomas\n",
    "    GROUP BY id_oferta),\n",
    "PrimeraCiudad AS (\n",
    "    SELECT \n",
    "        id_oferta,\n",
    "        MIN(ciudad) AS PrimeraCiudad\n",
    "    FROM ciudades\n",
    "    GROUP BY id_oferta)\n",
    "SELECT \n",
    "    (o.salario_desde + o.salario_hasta) / 2 AS salario,\n",
    "    o.variabe,\n",
    "    o.vacaciones,\n",
    "    o.teletrabajo,\n",
    "    o.jornada_tipo,\n",
    "    o.funciones,\n",
    "    o.contrato_tipo,\n",
    "    o.nivel_profesional,\n",
    "    o.formacion_minima,\n",
    "    o.personas_a_cargo,\n",
    "    o.experiencia,\n",
    "    pi.PrimerIdioma,\n",
    "    pc.PrimeraCiudad,\n",
    "    t.Tecnologia_1,\n",
    "    t.Tecnologia_2,\n",
    "    t.Tecnologia_3,\n",
    "    t.Tecnologia_4,\n",
    "    t.Tecnologia_5,\n",
    "    t.Tecnologia_6,\n",
    "    t.Tecnologia_7,\n",
    "    t.Tecnologia_8,\n",
    "    t.Tecnologia_9,\n",
    "    t.Tecnologia_10,\n",
    "    h.Habilidad_1,\n",
    "    h.Habilidad_2,\n",
    "    h.Habilidad_3,\n",
    "    h.Habilidad_4,\n",
    "    h.Habilidad_5\n",
    "FROM \n",
    "    ofertas o\n",
    "LEFT JOIN \n",
    "    PrimerIdioma pi ON o.id_oferta = pi.id_oferta\n",
    "LEFT JOIN \n",
    "    PrimeraCiudad pc ON o.id_oferta = pc.id_oferta\n",
    "LEFT JOIN \n",
    "    TecnologiasOferta t ON o.id_oferta = t.id_oferta\n",
    "LEFT JOIN \n",
    "    HabilidadesOferta h ON o.id_oferta = h.id_oferta;\n",
    "\"\"\"\n",
    "query_top10tec = \"\"\"\n",
    "SELECT t.tecnologia\n",
    "    FROM tecnologias_relacion tr\n",
    "    JOIN tecnologias t ON tr.tec_id = t.tec_id\n",
    "    GROUP BY tr.tec_id\n",
    "    ORDER BY COUNT(*) DESC\n",
    "    LIMIT 10;\n",
    "\"\"\"\n",
    "query_top5hab = \"\"\"\n",
    "    SELECT h.habilidad\n",
    "    FROM habilidades_relacion hr\n",
    "    JOIN habilidades h ON hr.hab_id = h.hab_id\n",
    "    GROUP BY hr.hab_id\n",
    "    ORDER BY COUNT(*) DESC\n",
    "    LIMIT 5;\n",
    "\"\"\"\n",
    "df_general = pd.read_sql(query, db)\n",
    "\n",
    "top10tec_df = pd.read_sql(query_top10tec, db)\n",
    "top10tec = top10tec_df[\"tecnologia\"].tolist()\n",
    "top5hab_df = pd.read_sql(query_top5hab, db)\n",
    "top5hab = top5hab_df[\"habilidad\"].tolist()\n",
    "\n",
    "db.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ae0e7ceb-5533-4892-89b9-1ad90562fb4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Renombramos columnas.\n",
    "\n",
    "df_general.rename(columns={\"Tecnologia_1\": top10tec[0]}, inplace=True)\n",
    "df_general.rename(columns={\"Tecnologia_2\": top10tec[1]}, inplace=True)\n",
    "df_general.rename(columns={\"Tecnologia_3\": top10tec[2]}, inplace=True)\n",
    "df_general.rename(columns={\"Tecnologia_4\": top10tec[3]}, inplace=True)\n",
    "df_general.rename(columns={\"Tecnologia_5\": top10tec[4]}, inplace=True)\n",
    "df_general.rename(columns={\"Tecnologia_6\": top10tec[5]}, inplace=True)\n",
    "df_general.rename(columns={\"Tecnologia_7\": top10tec[6]}, inplace=True)\n",
    "df_general.rename(columns={\"Tecnologia_8\": top10tec[7]}, inplace=True)\n",
    "df_general.rename(columns={\"Tecnologia_9\": top10tec[8]}, inplace=True)\n",
    "df_general.rename(columns={\"Tecnologia_10\": top10tec[9]}, inplace=True)\n",
    "df_general.rename(columns={\"Habilidad_1\": top5hab[0]}, inplace=True)\n",
    "df_general.rename(columns={\"Habilidad_2\": top5hab[1]}, inplace=True)\n",
    "df_general.rename(columns={\"Habilidad_3\": top5hab[2]}, inplace=True)\n",
    "df_general.rename(columns={\"Habilidad_4\": top5hab[3]}, inplace=True)\n",
    "df_general.rename(columns={\"Habilidad_5\": top5hab[4]}, inplace=True)\n",
    "df_general.rename(columns={\"PrimerIdioma\": \"idioma\"}, inplace=True)\n",
    "df_general.rename(columns={\"PrimeraCiudad\": \"ciudad\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b6e3bf68-2301-4c3b-8654-0da2bb71dbc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_general = df_general.dropna(subset=[\"salario\"]) #Eliminamos las columnas nan en salario para esta regresión."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c455cfea-d948-4f98-9fa3-bb50f780d92b",
   "metadata": {},
   "outputs": [],
   "source": [
    "encoder = {}\n",
    "encoder[\"salario\"] = None\n",
    "encoder[\"variable\"] = None\n",
    "encoder[\"vacaciones\"] = df_general[\"vacaciones\"].mode()[0]  # Moda de \"vacaciones\"\n",
    "encoder[\"teletrabajo\"] = {\"Presencial\": 0, \n",
    "                          \"Hibrido\": 1, \n",
    "                          \"Remoto\": 2}\n",
    "encoder[\"jornada_tipo\"] = {\"Por Horas\": 1,\n",
    "                           \"Media Jornada\": 2,\n",
    "                           \"Turno Rotatorio\": 3,\n",
    "                           \"Intensiva Tarde\": 4,\n",
    "                           \"Jornada Completa\": 4.1,\n",
    "                           \"Intensiva Mañana\": 4.3,\n",
    "                           \"Jornada Completa Flexible\": 4.4}\n",
    "encoder[\"funciones\"] = {cat: i for i, cat in enumerate(df_general[\"funciones\"].dropna().unique())}\n",
    "encoder[\"contrato_tipo\"] = {\"A determinar\": 1,\n",
    "                            \"Prácticas\": 2,\n",
    "                            \"Obra o servicio\": 3,\n",
    "                            \"Temporal\": 4,\n",
    "                            \"Freelance/Autónomo\": 5,\n",
    "                            \"Indefinido\": 6}\n",
    "encoder[\"nivel_profesional\"] = {\"Prácticas / Beca\": 1,\n",
    "                                \"Empleado\": 5,\n",
    "                                \"Mando Intermedio\": 20,\n",
    "                                \"Especialista\": 30,\n",
    "                                \"Director / Gerente\": 40,\n",
    "                                \"Administrador / Consejero\": 50}\n",
    "encoder[\"formacion_minima\"] = {\"Sin estudios\": 15,\n",
    "                               \"E.S.O. (Educación Secundaria Obligatoria)\": 16,\n",
    "                               \"FP1\": 18,\n",
    "                               \"Grado Medio\": 18,\n",
    "                               \"Bachillerato/COU\": 18.5,\n",
    "                               \"Certificado de Profesionalidad\": 19,\n",
    "                               \"Otros títulos, certificaciones y carnets\": 19.4,\n",
    "                               \"Otra Formación Tecnológica\": 19.8,\n",
    "                               \"FP2/Grado Superior\": 20,\n",
    "                               \"Diplomado\": 21,\n",
    "                               \"Ingeniero Técnico\": 21.5,\n",
    "                               \"Grado EEES (Bolonia)\": 22,\n",
    "                               \"Licenciado\": 23,\n",
    "                               \"Ingeniero Superior\": 23.5,\n",
    "                               \"Postgrado EEES (Máster)\": 24,\n",
    "                               \"Doctorado\": 28}\n",
    "encoder[\"personas_a_cargo\"] = 0  # Se rellena con 0 si hay valores nulos\n",
    "encoder[\"experiencia\"] = {\"Sin Experiencia\": 1,\n",
    "                          \"Menos de un año\": 2,\n",
    "                          \"1 año\": 2,\"2 años\": 5,\n",
    "                          \"3 años\": 5,\"3-5 años\": 5,\n",
    "                          \"Más de 5 años\": 5,\n",
    "                          \"Más de 10 años\": 10,\n",
    "                          \"5 años o más\": 10}\n",
    "encoder[\"idioma\"] = {cat: i for i, cat in enumerate(df_general[\"idioma\"].dropna().unique())}\n",
    "encoder[\"ciudad\"] = {cat: i for i, cat in enumerate(df_general[\"ciudad\"].dropna().unique())}\n",
    "encoder[\"sap ewm\"] = None \n",
    "encoder[\"python\"] = None \n",
    "encoder[\"sql\"] = None \n",
    "encoder[\"AWS\"] = None \n",
    "encoder[\"JavaScript\"] = None \n",
    "encoder[\"Python\"] = None \n",
    "encoder[\"peoplesoft\"] = None \n",
    "encoder[\"Docker\"] = None \n",
    "encoder[\"aws\"] = None \n",
    "encoder[\"Programador\"] = None \n",
    "encoder[\"Proactividad\"] = None \n",
    "encoder[\"Trabajo en equipo\"] = None \n",
    "encoder[\"Aprendizaje Continuo\"] = None \n",
    "encoder[\"Capacidad de autogestión\"] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9d048364-974c-48ec-b060-8e8547fb3874",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_50378/1148020815.py:10: FutureWarning: Series.replace without 'value' and with non-dict-like 'to_replace' is deprecated and will raise in a future version. Explicitly specify the new values instead.\n",
      "  return df[column_name].replace(mapping).fillna(df[column_name])\n",
      "/tmp/ipykernel_50378/1148020815.py:10: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  return df[column_name].replace(mapping).fillna(df[column_name])\n",
      "/tmp/ipykernel_50378/1148020815.py:10: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  return df[column_name].replace(mapping).fillna(df[column_name])\n"
     ]
    }
   ],
   "source": [
    "#Usamos el encoder para codificar el df\n",
    "\n",
    "df_codificado = df_general.copy()\n",
    "def transform_column(df, column_name, encoder):\n",
    "    if column_name in encoder:\n",
    "        mapping = encoder[column_name]\n",
    "        if mapping is None:\n",
    "            return df[column_name]\n",
    "        else:\n",
    "            return df[column_name].replace(mapping).fillna(df[column_name])\n",
    "    else:\n",
    "        return df[column_name]\n",
    "for column in df_codificado.columns:\n",
    "    df_codificado[column] = transform_column(df_codificado, column, encoder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7d35aa82-d3dc-4967-92c3-d119906d0dd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Escalamos el df con MinMaxScaler\n",
    "\n",
    "scaler = MinMaxScaler()\n",
    "df_scaled = scaler.fit_transform(df_codificado)  # Esto devuelve un numpy array\n",
    "# Convertir el resultado a un df\n",
    "df_scaled = pd.DataFrame(df_scaled, columns=df_codificado.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b2a46bb5-42a9-4017-a966-1362f73b3a27",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Imputamos valores faltantes con KNNImputer\n",
    "\n",
    "imputer = KNNImputer(n_neighbors=5)\n",
    "columnas = df_scaled.columns\n",
    "df_imputado = imputer.fit_transform(df_scaled)\n",
    "df_imputado = pd.DataFrame(df_imputado, columns = df_scaled.columns)\n",
    "df_imputado.to_csv(\"df_imputado_regresion_salario\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6dbc01aa-5b4d-4056-8b0c-f1e0e3ffeba7",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/__init__.py\", line 73, in <module>\n",
      "    from .base import clone  # noqa: E402\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/base.py\", line 19, in <module>\n",
      "    from .utils._estimator_html_repr import _HTMLDocumentationLinkMixin, estimator_html_repr\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/__init__.py\", line 15, in <module>\n",
      "    from ._chunking import gen_batches, gen_even_slices\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/_chunking.py\", line 11, in <module>\n",
      "    from ._param_validation import Interval, validate_params\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/_param_validation.py\", line 17, in <module>\n",
      "    from .validation import _is_arraylike_not_scalar\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/validation.py\", line 21, in <module>\n",
      "    from ..utils._array_api import _asarray_with_order, _is_numpy_namespace, get_namespace\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/_array_api.py\", line 17, in <module>\n",
      "    from .fixes import parse_version\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/fixes.py\", line 20, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 26, in <module>\n",
      "    from pandas.compat import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/compat/__init__.py\", line 27, in <module>\n",
      "    from pandas.compat.pyarrow import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/compat/pyarrow.py\", line 8, in <module>\n",
      "    import pyarrow as pa\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "    raise ImportError(msg)\n",
      "ImportError: \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/__init__.py\", line 73, in <module>\n",
      "    from .base import clone  # noqa: E402\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/base.py\", line 19, in <module>\n",
      "    from .utils._estimator_html_repr import _HTMLDocumentationLinkMixin, estimator_html_repr\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/__init__.py\", line 15, in <module>\n",
      "    from ._chunking import gen_batches, gen_even_slices\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/_chunking.py\", line 11, in <module>\n",
      "    from ._param_validation import Interval, validate_params\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/_param_validation.py\", line 17, in <module>\n",
      "    from .validation import _is_arraylike_not_scalar\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/validation.py\", line 21, in <module>\n",
      "    from ..utils._array_api import _asarray_with_order, _is_numpy_namespace, get_namespace\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/_array_api.py\", line 17, in <module>\n",
      "    from .fixes import parse_version\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/fixes.py\", line 20, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 49, in <module>\n",
      "    from pandas.core.api import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/core/api.py\", line 9, in <module>\n",
      "    from pandas.core.dtypes.dtypes import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/core/dtypes/dtypes.py\", line 24, in <module>\n",
      "    from pandas._libs import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "    raise ImportError(msg)\n",
      "ImportError: \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/__init__.py\", line 73, in <module>\n",
      "    from .base import clone  # noqa: E402\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/base.py\", line 19, in <module>\n",
      "    from .utils._estimator_html_repr import _HTMLDocumentationLinkMixin, estimator_html_repr\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/__init__.py\", line 15, in <module>\n",
      "    from ._chunking import gen_batches, gen_even_slices\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/_chunking.py\", line 11, in <module>\n",
      "    from ._param_validation import Interval, validate_params\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/_param_validation.py\", line 17, in <module>\n",
      "    from .validation import _is_arraylike_not_scalar\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/validation.py\", line 21, in <module>\n",
      "    from ..utils._array_api import _asarray_with_order, _is_numpy_namespace, get_namespace\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/_array_api.py\", line 17, in <module>\n",
      "    from .fixes import parse_version\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/fixes.py\", line 20, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 26, in <module>\n",
      "    from pandas.compat import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/compat/__init__.py\", line 27, in <module>\n",
      "    from pandas.compat.pyarrow import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/compat/pyarrow.py\", line 8, in <module>\n",
      "    import pyarrow as pa\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "    raise ImportError(msg)\n",
      "ImportError: \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n",
      "\n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "Traceback (most recent call last):  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/backend/popen_loky_posix.py\", line 180, in <module>\n",
      "    exitcode = process_obj._bootstrap()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/multiprocessing/process.py\", line 314, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/joblib/externals/loky/process_executor.py\", line 426, in _process_worker\n",
      "    call_item = call_queue.get(block=True, timeout=timeout)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/multiprocessing/queues.py\", line 122, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/__init__.py\", line 73, in <module>\n",
      "    from .base import clone  # noqa: E402\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/base.py\", line 19, in <module>\n",
      "    from .utils._estimator_html_repr import _HTMLDocumentationLinkMixin, estimator_html_repr\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/__init__.py\", line 15, in <module>\n",
      "    from ._chunking import gen_batches, gen_even_slices\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/_chunking.py\", line 11, in <module>\n",
      "    from ._param_validation import Interval, validate_params\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/_param_validation.py\", line 17, in <module>\n",
      "    from .validation import _is_arraylike_not_scalar\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/validation.py\", line 21, in <module>\n",
      "    from ..utils._array_api import _asarray_with_order, _is_numpy_namespace, get_namespace\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/_array_api.py\", line 17, in <module>\n",
      "    from .fixes import parse_version\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/sklearn/utils/fixes.py\", line 20, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/__init__.py\", line 49, in <module>\n",
      "    from pandas.core.api import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/core/api.py\", line 9, in <module>\n",
      "    from pandas.core.dtypes.dtypes import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pandas/core/dtypes/dtypes.py\", line 24, in <module>\n",
      "    from pandas._libs import (\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/pyarrow/__init__.py\", line 65, in <module>\n",
      "    import pyarrow.lib as _lib\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/bosser/anaconda3/lib/python3.12/site-packages/numpy/core/_multiarray_umath.py\", line 44, in __getattr__\n",
      "    raise ImportError(msg)\n",
      "ImportError: \n",
      "A module that was compiled using NumPy 1.x cannot be run in\n",
      "NumPy 2.2.3 as it may crash. To support both 1.x and 2.x\n",
      "versions of NumPy, modules must be compiled with NumPy 2.0.\n",
      "Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.\n",
      "\n",
      "If you are a user of the module, the easiest solution will be to\n",
      "downgrade to 'numpy<2' or try to upgrade the affected module.\n",
      "We expect that some modules will need time to support NumPy 2.\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mejores hiperparámetros: {'max_depth': 20, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 300}\n",
      "MSE con el mejor modelo: 0.002676331707909604\n"
     ]
    }
   ],
   "source": [
    "#Con la grilla de hiperparámetros, escogemos el mejor modelo posible.\n",
    "\n",
    "X = df_imputado.drop(\"salario\", axis=1)\n",
    "y = df_imputado[\"salario\"] #Separamos la columna objetivo\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15, random_state=42)\n",
    "\n",
    "model = RandomForestRegressor(random_state=42)\n",
    "\n",
    "# Grilla de hiperparámetros\n",
    "param_grid = {\n",
    "    \"n_estimators\": [100, 200, 300],\n",
    "    \"max_depth\": [None, 10, 20, 30],\n",
    "    \"min_samples_split\": [2, 5, 10],\n",
    "    \"min_samples_leaf\": [1, 2, 4],\n",
    "    \"max_features\": [\"sqrt\", \"log2\"]}\n",
    "\n",
    "# Configurar Grid Search\n",
    "grid_search = GridSearchCV(estimator=model,\n",
    "                           param_grid=param_grid,\n",
    "                           scoring=\"neg_mean_squared_error\",\n",
    "                           cv=5,\n",
    "                           n_jobs=-1)\n",
    "\n",
    "# Ajustar Grid Search a los datos de entrenamiento\n",
    "grid_search.fit(X_train, y_train)\n",
    "\n",
    "# Obtener los mejores hiperparámetros\n",
    "best_params = grid_search.best_params_\n",
    "print(\"Mejores hiperparámetros:\", best_params)\n",
    "\n",
    "# Evaluar el mejor modelo en el conjunto de prueba\n",
    "best_model = grid_search.best_estimator_\n",
    "y_pred = best_model.predict(X_test)\n",
    "mse = mean_squared_error(y_test, y_pred)\n",
    "print(f\"MSE con el mejor modelo: {mse}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2e2f0c99-9f5b-4dfa-b0a9-01cad0313553",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['transformadores_y_modelo_regresion_salario.pkl']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Guardamos encoder, scaler, imputer y modelo.\n",
    "\n",
    "transformadores = {\"encoder\": encoder,\n",
    "                   \"scaler\": scaler,\n",
    "                   \"imputer\": imputer,\n",
    "                   \"model\": best_model}\n",
    "joblib.dump(transformadores, \"transformadores_y_modelo_regresion_salario.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bca5aa18-6a3f-4f8d-8375-5e635b225dbc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0074b04e-a1d1-48a4-936d-8da74a6e19e7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29897edd-eccd-48e5-ae14-cdb48df4102b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6242489-9630-4765-8856-0e44c2b42e93",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fc54615-7084-42a5-a4d1-4c37e1f9a7f8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e93f5022-40cd-4de4-8203-f5dbdf31dc37",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3034338-6db1-4d9a-aef8-73bf5ce59f4c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "660ba113-1990-4211-bb8b-a34f623035fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e34cdd58-683d-4e04-8ef9-43f6ad4de140",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e476d15e-d376-482c-840d-ae5e410cd519",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd39deba-0a39-463b-9a6b-865363f31be3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a94735f1-2ec9-453b-a9f2-887db880563a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bcfc9dc-f634-456f-b792-17e83913ac75",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ade645b5-4fad-4d67-a73f-025c710f88a7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
